File: src/Bully.java
import org.apache.xmlrpc.XmlRpcException;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashSet;
import java.util.Vector;

class Bully {


    public static boolean startElection(int ownPort, HashSet<String> connectedNodes) {
        String response;

        System.out.println("node " + ownPort + " starts election process!");
        for (String node : connectedNodes) {
            if(extractPortFromIPnPort(node) <= ownPort) continue;
            response = messageNode(ownPort, node);

            switch (response) {
                case "Continue" :
                    //TODO: implement
                    break;
                case "Stop" :
                    System.out.println(node + " is taking over the election process");
                    return false;
                case "Lost" :
                    signOffDisconnectedNode(node);
                    break;
                default :
                    System.err.println("Something went wrong!");
            }
        }
        return true;
    }

    private static String messageNode(int ownPort, String node) {

        Vector<Object> params = new Vector<>();
        System.out.println("sending election message to: " + node);
        try {
            Client.config.setServerURL(new URL(Client.getFullAddress(Client.urlFormatter(node))));
            Client.xmlRpcClient.setConfig(Client.config);
            params.removeAllElements();
            params.add(ownPort);

            return (String) Client.xmlRpcClient.execute("Node.rpcElectionRequest", params);

        } catch (XmlRpcException e) {
            e.printStackTrace();
            return "Lost";
        } catch (MalformedURLException e) {
            e.printStackTrace();
            return "Error";
        }
    }

    private static boolean signOffDisconnectedNode(String node) {
        Vector<Object> params = new Vector<>();

        try {
            Client.config.setServerURL(new URL(Client.getFullAddress(Client.urlFormatter(Client.nodeIPnPort))));

            Client.xmlRpcClient.setConfig(Client.config);
            params.removeAllElements();
            params.add(node);

            return (boolean) Client.xmlRpcClient.execute("Node.signOff", params);
        } catch (MalformedURLException e) {
            e.printStackTrace();
            return false;
        } catch (XmlRpcException e) {
            e.printStackTrace();
            return false;
        }
    }

    private static int extractPortFromIPnPort(String nodeIPnPort) {
        int port;
        String portString = nodeIPnPort.substring(nodeIPnPort.indexOf(":")+1);
        port = Integer.parseInt(portString);

        return port;
    }

}

File: c++/DSProject/Bully.cpp
#include "Bully.h"

Bully::Bully() : QObject(){
	
}

bool Bully::startElection(int ownPort, QList<QVariant> connectedNodes) {
	QString response;

    std::cout << std::endl << "node " << ownPort << " starts election process!" << std::endl;
    foreach(QVariant node, connectedNodes) {
        if (extractPortFromIPnPort(node.toString()) <= ownPort) continue;
		response = messageNode(ownPort, node.toString());

		if (response == "Continue"){
			//TODO: implement
		}
		else if (response == "Stop"){
			std::cout << std::endl << "[Bully] " << node.toString().toStdString() << " is taking over the election process";
			return false;
		}
		else if (response == "Lost"){
			std::cout << std::endl << "[Bully] " << node.toString().toStdString() << " is not answering therefore will be disconnected";
			signOffDisconnectedNode(node.toString());
		}
		else
            std::cout << std::endl << "Something went wrong!";
	}
	return true;
}

QString Bully::messageNode(int ownPort, QString node) {

	QList<QVariant> params;
    std::cout << std::endl << "sending election message to: " << node.toStdString();
	Client* xmlRpcClient = new Client();
    xmlRpcClient->setHostAndPort(node);
	params.append(ownPort);

    xmlRpcClient->execute("rpcElectionRequest", params);

    while (xmlRpcClient->isWaiting())
        QCoreApplication::processEvents();

    QString response = xmlRpcClient->response().toString();

	if (response.isEmpty() || response.isNull()) {
		return "Lost";
	}

	return response;
}

bool Bully::signOffDisconnectedNode(QString node) {
	QList<QVariant> params;

	Client* xmlRpcClient = new Client();
    xmlRpcClient->setHostAndPort(Client::nodeIPnPort);
	params.append(node);

    xmlRpcClient->execute("signOff", params);
    while (xmlRpcClient->isWaiting())
        QCoreApplication::processEvents();

    return xmlRpcClient->response().toBool();
}

int Bully::extractPortFromIPnPort(QString nodeIPnPort) {
	return nodeIPnPort.right(nodeIPnPort.length() - nodeIPnPort.indexOf(":") - 1).toInt();
}
